server/src/main/java/org/elasticsearch/transport/InboundAggregator.java

```diff
@@ -0,0 +1,128 @@
+
+    private ReleasableBytesReference firstContent;
+    private ArrayList<ReleasableBytesReference> contentAggregation;
+    private Header currentHeader;
+
+    public void headerReceived(Header header) {
+        if (currentHeader != null) {
+            currentHeader = null;
+            throw new IllegalStateException("Header already received.");
+        }
+
+        currentHeader = header;
+    }
+
+    public void aggregate(ReleasableBytesReference content) {
+        if (currentHeader == null) {
+            content.close();
+            throw new IllegalStateException("Received content without header");
+        } else {
+            if (isFirstContent()) {
+                firstContent = content.retain();
+            } else {
+                if (contentAggregation == null) {
+                    contentAggregation = new ArrayList<>(4);
+                    contentAggregation.add(firstContent);
+                    firstContent = null;
+                }
+                contentAggregation.add(content.retain());
+            }
+        }
+    }
+
+    public Header cancelAggregation() {
+        if (currentHeader == null) {
+            throw new IllegalStateException("Aggregation cancelled, but no aggregation had begun");
+        } else {
+            final Header header = this.currentHeader;
+            closeAggregation();
+            return header;
+        }
+    }
+
+    public InboundMessage finishAggregation() throws IOException {
+        final ReleasableBytesReference releasableContent;
+        if (isFirstContent()) {
+            releasableContent = ReleasableBytesReference.wrap(BytesArray.EMPTY);
+        } else if (contentAggregation == null) {
+            releasableContent = firstContent;
+        } else {
+            final ReleasableBytesReference[] references = contentAggregation.toArray(new ReleasableBytesReference[0]);
+            final CompositeBytesReference content = new CompositeBytesReference(references);
+            releasableContent = new ReleasableBytesReference(content, () -> Releasables.close(references));
+        }
+        final InboundMessage aggregated = new InboundMessage(currentHeader, releasableContent);
+        clearAggregation();
+        boolean success = false;
+        try {
+            if (aggregated.getHeader().needsToReadVariableHeader()) {
+                aggregated.getHeader().finishParsingHeader(aggregated.openOrGetStreamInput());
+            }
+            success = true;
+            return aggregated;
+        } finally {
+            if (success == false) {
+                aggregated.close();
+            }
+        }
+    }
+
+    public boolean isAggregating() {
+        return currentHeader != null;
+    }
+
+    private boolean isFirstContent() {
+        return firstContent == null && contentAggregation == null;
+    }
+
+    @Override
+    public void close() {
+        closeAggregation();
+    }
+
+    private void closeAggregation() {
```

Comment:
I made a few changes in this area. I want close to be called a single time and the aggregator cannot be reused after the close call (to prevent leaks from a theoretical reentrancy bug.)